### PR TITLE
[XdsClient] fix edge case when rapidly subscribing and unsubscribing

### PR DESCRIPTION
**Pull request by markdroth**

### CODE REVIEW

File: test/core/xds/xds_client_test.cc [hunk 1]
```diff
@@ -4708,6 +4717,209 @@ TEST_F(XdsClientTest, DoNotSendDoesNotExistForCachedResource) {
...
+  // XdsClient should have sent an ACK message to the xDS server.
+  request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"1", /*response_nonce=*/"A",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  stream->CompleteSendMessageFromClient();
+  // Start a watch for a second resource.
+  auto watcher2 = StartFooWatch("foo2");
```
Comment (pawbhard) [inline test/core/xds/xds_client_test.cc:4781]: Do we need foo2 in this testcase ? I think we can cleanup second resource from this testcase

File: src/core/xds/xds_client/xds_client.cc [hunk 2]
```diff
@@ -737,6 +740,7 @@ XdsClient::XdsChannel::AdsCall::AdsCall(
 }
 
 void XdsClient::XdsChannel::AdsCall::Orphan() {
+  xds_client()->MaybeRemoveUnsubscribedCacheEntriesLocked(xds_channel());
```
Comment (pawbhard) [inline src/core/xds/xds_client/xds_client.cc:743]: We can check if this is needed

**Review by pawbhard — APPROVED** [review]
LGTM

File: src/core/xds/xds_client/xds_client.cc [hunk 3]
```diff
@@ -1769,60 +1769,68 @@ void XdsClient::CancelResourceWatch(const XdsResourceType* type,
...
-        resource_map.erase(resource_it++);
-      } else {
-        ++resource_it;
+  for (auto authority_it = authority_state_map_.begin();
+       authority_it != authority_state_map_.end();) {
+    AuthorityState& authority_state = authority_it->second;
+    if (authority_state.xds_channels.back() == xds_channel) {
+      // Find type map.
+      auto type_it = authority_state.type_map.find(type);
+      if (type_it == authority_state.type_map.end()) return;
```
Comment (pawbhard) [inline src/core/xds/xds_client/xds_client.cc:1815]: Reason for return rather than continue to next iteration

Comment (markdroth) [reply src/core/xds/xds_client/xds_client.cc:1815]: Good catch!  Fixed.

**Comment by markdroth** [comment]
The tests showed a ubsan problem, which caused me to rethink the approach a bit, and I found a simpler way to address it.  PTAL.

**Comment by markdroth** [comment]
Looks like this is triggering a bug in the xDS-enabled server code.  I'll debug it more on Monday.

**Comment by markdroth** [comment]
Okay, I think this should be ready for another round of review now.  Hopefully, the tests won't show any new problems.

**Comment by markdroth** [comment]
I just realized that there's still one more edge case here I need to handle.  Please hold off on reviewing while I figure this out...

**Comment by markdroth** [comment]
Okay, I think I'm finally happy with this.  PTAL.

**Review by pawbhard — APPROVED** [review]
LGTM
